src/loader.ts

```diff
@@ -1,207 +1,101 @@
-  const { protocol } = new URL(modulePath);
-  const isNode = protocol === "node:";
-  const isFile = protocol === "file:";
-
-  if (isNode || isFile) return specifier;
-
-  const cachePath = constructCachePath({ cache, modulePath })
-
-  cacheMap.set(`file://${cachePath}`, modulePath);
-  if (existsSync(cachePath)) return cachePath
-  const code = await (await fetch(modulePath).then(response => {
-    if (!response.ok) {
-      throw Error(`404: Module not found: ${modulePath}`);
-    }
-    return response;
-  })).text();
-
-  ensureFileSync(cachePath);
-  writeFileSync(cachePath, code);
-
-  return cachePath;
+export const resolve = async (
+  specifier: string,
+  { parentURL }: Context,
+  nextResolve: NextResolve,
+  options: ResolveOptions = config?.values,
+) => {
+  const { basePath, cachePath, debug: isDebugging = false, importmapPath, cacheMap = initialCacheMap } = options || {};
+  log.setLogger(isDebugging);
+  try {
+    // define importmap path
+    const cwd = process.cwd();
+    const pathToImportMap = importmapPath || constructUrlPath(basePath, cwd, isDebugging);
+    const nodeImportMapPath = constructPath("node.importmap", pathToImportMap);
+    log.debug("resolve:nodeImportMapPath:", { cwd, pathToImportMap, nodeImportMapPath });
+    if (!nodeImportMapPath) throw new Error("Failed in resolving import map path");
+
+    // define cache path
+    const pathToCache = cachePath || parentURL;
+    log.debug("resolve:pathToCache:", { pathToCache });
+    if (!pathToCache) throw new Error("Failed in resolving cache path");
+
+    // construct importmap
+    const importmap = constructImportMap(nodeImportMapPath);
+    log.debug("resolve:importmap:", { importmap });
+    if (!importmap) throw new Error("Failed in constructing import map");
+
+    // construct cache map path
+    const cacheMapPath = cacheMap.get(pathToCache);
+    log.debug("resolve:cacheMapPath:", { cacheMapPath });
+    if (!cacheMapPath) throw new Error("Failed in resolving cache map path");
+
+    // construct module path
+    const modulePath = importmap.resolve(specifier, cacheMapPath);
+    log.debug("resolve:modulePath:", { modulePath });
+    if (!modulePath) throw new Error("Failed in resolving module path");
+
+    // resolve URL
+    const { protocol = "" } = new URL(modulePath);
+    const isNode = protocol === "node:";
+    const isFile = protocol === "file:";
+    log.debug("resolve:protocol:", { protocol, isNode, isFile });
+    if (isNode || isFile) throw new Error("Failed in resolving URL");
+
+    // get node module information
+    const moduleMetaData = await parseUrlPkg(modulePath);
+    log.debug("resolve:moduleMetaData:", { moduleMetaData });
+    if (!moduleMetaData) throw new Error("Failed in parsing module meta data");
+
+    // construct node module cache path
+    const {
+      pkg: { name, version },
+    } = moduleMetaData;
+    const nodeModuleCachePath = constructPath(`${name}@${version}`, pathToCache);
```

Comment:
I remember we are using this to maintain the support for specific package of specific version. While i was working on this. I just realised we can actually use `new URL(modulePath)` to get the `route` path. And then it contains the `npm:<package-name>@<version>` data to preserve unique folders.

So, we don't need to run this `parseUrlPkg` it's just a hunch that we can test it out. Since, this just removes the dependency on the `@jspm/generator` package. And helps in less maintenance as we don't need to maintain all the updates when the generator is updates.

Reply:
This comments competes [with another comment](https://github.com/jspm/node-importmap-http-loader/pull/8/files/4a8d90033b3d0cc82a953ddf9f3f747d6f1ba478#r1208065631). So I'll let that be resolved in another PR 👍